### ISSUE TITLE
False positive on namespace exports

### ISSUE THREAD

**Issue by Faithfinder** [webpro/knip#475]
[Repro](https://codesandbox.io/p/devbox/knip-unused-namespace-exports-repro-jlnq5m?file=%2Fsrc%2Findex.ts%3A6%2C4&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clrsasof30006356n1g7nu728%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clrsasof30002356nsi7dek6r%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clrsasof30004356n6vokpb8f%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clrsasof30005356ngf11sj0n%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clrsasof30002356nsi7dek6r%2522%253A%257B%2522id%2522%253A%2522clrsasof30002356nsi7dek6r%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clrtcuzg6002e356mckft4e2v%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A12%252C%2522startColumn%2522%253A27%252C%2522endLineNumber%2522%253A12%252C%2522endColumn%2522%253A27%257D%255D%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clrtefnl80002356md4op0new%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A6%252C%2522startColumn%2522%253A4%252C%2522endLineNumber%2522%253A6%252C%2522endColumn%2522%253A4%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.ts%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clrtefnl80002356md4op0new%2522%257D%252C%2522clrsasof30005356ngf11sj0n%2522%253A%257B%2522id%2522%253A%2522clrsasof30005356ngf11sj0n%2522%252C%2522activeTabId%2522%253A%2522clrtef0g201q6356mlhr7a2yn%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522SANDBOX_INFO%2522%252C%2522id%2522%253A%2522clrtef0g201q6356mlhr7a2yn%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%252C%2522clrsasof30004356n6vokpb8f%2522%253A%257B%2522id%2522%253A%2522clrsasof30004356n6vokpb8f%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clrsasof30003356n681x71dt%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%257D%252C%257B%2522id%2522%253A%2522clrtczm7t00tl356mvtwv9daz%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clrtczm1y0008eege9lqmcrfv%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clrsasof30003356n681x71dt%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D).

When the whole namespace is assigned to an object property, knip thinks nothing is used, when actually, everything is used

**Comment by webpro** [comment]
> when actually, everything is used

This is exactly why there's a separate category for "Unused exports in namespaces" (in addition to "Unused exports"), since, to be fair, Knip has no way to know if "everything is used".

One could argue you should use the exports explicitly:

```ts
import { one, two } from './star';

function usage() {
  const hello = {
    one,
    two
  }
}
```

One could also argue that by seeing e.g. `* as stuff`, Knip should mark all exports attached to `stuff` as used. Yet that also has the disadvantage of missing out on actually unused exports from that module.

This is why currently I'd recommend to explicitly import the individual exports. This may not be convenient at times, but I'm not sure yet what's the best approach here for those cases.

**Comment by Faithfinder** [comment]
Hmm, I wonder if this should be it's own category, "Ambigous imports" or something? "Because of the way your code is written, we can't tell what's used and what's not. We suggest you rewrite, or turn this category off if you don't care". 

I can't call this _unused_, that's misleading and led me to blink at the screen a lot when stuff went wrong after I deleted the supposedly unused things. IMO, false positive is worse than false negative in case of unused code, because it can lead to wild goose chases.

I also don't want to turn this category off entirely, because it did find some _entirely_ unused things.

**Comment by Faithfinder** [comment]
BTW, to be entirely clear, the two cases that lit up in our codebase were both related to package setups.

One was testing-library/react, set up according to their [`custom queries` example](https://testing-library.com/docs/react-testing-library/setup#add-custom-queries)

![image](https://github.com/webpro/knip/assets/25051063/7aa34fd1-34cb-458e-ba27-3b6b129ffde6)

Another was Nexus's [`makeSchema`](https://nexusjs.org/docs/api/make-schema#types), again with setup like the docs show

![image](https://github.com/webpro/knip/assets/25051063/7c72848b-0747-43d5-82ed-0d5efa955b40)

**Comment by webpro** [comment]
Thanks for pushing this @Faithfinder, I agree with basically everything.

I've given it some thought and just published a snapshot tagged `nsexport` on npm.

- Adds two new issue types "Exports in used namespace" (and "Exported types in used namespace").
- The types are not included by default.
- They can be included by using `--include nsExport,nsType`
- Handles most strict/obvious cases I can think of atm (what we've discussed in this issue and a few more), but still under investigation.
- Nothing is final yet (including names and titles), probably needs another iteration or two to make it good enough. Also no documentation yet.

Would be great if you could try it on your codebase(s) too. You could install using `npm install knip@nsexport` (or other package manager).

**Comment by Faithfinder** [comment]
Hey, BTW, I got downsized from my company on Monday, so, sorry, but no more feedback for you >.>

**Comment by webpro** [comment]
Sorry to hear this @Faithfinder! Thanks a lot for all your contributions here.

**Comment by webpro** [comment]
:rocket: _This issue has been resolved in v5.0.0. See [Release 5.0.0](https://github.com/webpro/knip/releases/tag/5.0.0) for release notes._
